File: app_python/app/domain/visits.py
from abc import ABC, abstractmethod
from pathlib import Path

from pydantic_settings import BaseSettings, SettingsConfigDict


class VisitsStorageConfig(BaseSettings):
    model_config = SettingsConfigDict(env_file=".env.visits")


class VisitsFileStorageConfig(VisitsStorageConfig):
    file_path: Path = Path("visits")
    encoding: str = "utf-8"


class VisitsStorage(ABC):
    @abstractmethod
    def __init__(self, config: VisitsStorageConfig): ...

    @abstractmethod
    async def read_data(self) -> int: ...

    @abstractmethod
    async def increment_data(self) -> None: ...


class VisitsFileStorage(VisitsStorage):
    def __init__(self, config: VisitsFileStorageConfig):
        self.file = config.file_path
        self._encoding = config.encoding

        self.file_mount()

    def file_mount(self):
        if self.file.exists():
            return

        with open(self.file, "w", encoding=self._encoding) as file:
            file.write(str(0))

    async def read_data(self) -> int:
        with open(self.file, "r", encoding=self._encoding) as file:
            return int(file.read())

    async def increment_data(self) -> None:
        with open(self.file, "r+", encoding=self._encoding) as file:
            visits = int(file.read() or "0")
            visits += 1

            file.seek(0)
            file.write(str(visits))

File: app_rust/app/src/service/visits.rs
use std::fs::OpenOptions;
use std::io::{Read, Seek, SeekFrom, Write};

fn file_mount(path: &String) -> std::io::Result<()> {
    if std::path::Path::new(path).exists() {
        return Ok(());
    }
    let mut file = OpenOptions::new().write(true).create_new(true).open(path)?;

    write!(file, "0")
}

pub struct VisitsFileStorage {
    file_path: String,
}

impl VisitsFileStorage {
    pub fn new(file_path_: String) -> Self {
        let _ = file_mount(&file_path_);
        Self {
            file_path: file_path_.clone(),
        }
    }

    pub fn read_data(&self) -> u64 {
        return std::fs::read_to_string(&self.file_path)
            .unwrap()
            .parse::<u64>()
            .unwrap();
    }

    pub fn increment(&self) -> std::io::Result<()> {
        let mut file = OpenOptions::new()
            .read(true)
            .write(true)
            .open(&self.file_path)?;

        let mut visits_number = String::new();
        file.read_to_string(&mut visits_number)?;

        let old_count: u64 = visits_number.trim().parse().unwrap();
        let new_count = old_count + 1;

        file.seek(SeekFrom::Start(0))?;

        write!(file, "{}", new_count)
    }
}
